pandas/core/indexes/base.py

```diff
@@ -4211,6 +4227,7 @@ def __getitem__(self, key):
         else:
             return result
 
+    @final
```

Comment:
```
from pandas.core.indexes.base import Index

reveal_type(Index()._can_hold_identifiers_and_holds_name(""))
```
on master
```
$ mypy pandas/test.py
pandas\test.py:3: note: Revealed type is 'builtins.bool'
(pandas-dev) 
```

this PR..
```
$ mypy pandas/test.py
pandas\_typing.py:22: error: Module 'typing' has no attribute 'final'
pandas\test.py:3: note: Revealed type is 'Any'
Found 1 error in 1 file (checked 1 source file)```

Reply:
There was a suggestion to type final as `def final(x: F) -> F`, but that failed because F is undefined.  Would restoring that solve this problem? and if so, what should F be?

Reply:
the link in the comment explains. add..
```
FuncType = Callable[..., Any]
F = TypeVar('F', bound=FuncType)
```